src/lib.rs

```diff
@@ -654,29 +658,68 @@ impl Maildir {
         // to new dir. see for example http://www.courier-mta.org/maildir.html.
         let pid = std::process::id();
         let hostname = gethostname::gethostname();
+        let hostname = hostname.to_string_lossy();
 
         // loop when conflicting filenames occur, as described at
         // http://www.courier-mta.org/maildir.html
         // this assumes that pid and hostname don't change.
-        let mut ts = time::SystemTime::now().duration_since(time::UNIX_EPOCH)?;
         let mut tmppath = self.path.clone();
         tmppath.push("tmp");
+
+        let mut file;
+        let mut secs;
+        let mut nanos;
+        let mut counter;
+
         loop {
-            tmppath.push(format!(
-                "{}.M{}P{}.{}",
-                ts.as_secs(),
-                ts.subsec_nanos(),
-                pid,
-                hostname.to_string_lossy()
-            ));
-            if !tmppath.exists() {
-                break;
+            let ts = time::SystemTime::now().duration_since(time::UNIX_EPOCH)?;
+            secs = ts.as_secs();
+            nanos = ts.subsec_nanos();
+            counter = COUNTER.fetch_add(1, Ordering::SeqCst);
+
+            tmppath.push(format!("{secs}.X{counter:x}M{nanos}P{pid}.{hostname}"));
```

Comment:
According to http://cr.yp.to/proto/maildir.html the `X` prefix is for boot number. Here you're using more like a sequence number so I'd prefer you used `#` instead of `X`.